### PR TITLE
Add DOS detection module

### PR DESCRIPTION
**Pull request by b-mueller**
This module returns a low severity warning when a state change or call is executed in a loop. It correctly detects the following SWC test cases:

https://smartcontractsecurity.github.io/SWC-registry/docs/SWC-128#dos-addresssol
https://smartcontractsecurity.github.io/SWC-registry/docs/SWC-128#dos-numbersol
https://smartcontractsecurity.github.io/SWC-registry/docs/SWC-128#dos-simplesol

### CODE REVIEW

File: mythril/analysis/modules/dos.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+        opcode = state.get_current_instruction()["opcode"]
+
+        if opcode == "JUMPI":
+
+            target = util.get_concrete_int(state.mstate.stack[-1])
+
+            try:
+                transaction = state.current_transaction
+                self._jumpdest_count[transaction]
```
Comment (norhh) [inline mythril/analysis/modules/dos.py]: we can use ` if transaction in self._jumpdest_count` in this case.

File: mythril/analysis/modules/dos.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+                try:
+                    if self._jumpdest_count[transaction][target] > 3:
+                        state.annotate(
```
Comment (norhh) [inline mythril/analysis/modules/dos.py]: I think we might be able to reduce the number of annotate()'s here, by checking if it's 4 then annotate and ++, >4 then ignore, might save us some memory.

File: mythril/analysis/modules/dos.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+            transaction = state.current_transaction
+            if state.current_transaction in self._jumpdest_count:
+
+                try:
+                    if self._jumpdest_count[transaction][target] == 4:
+                        state.annotate(
+                            LoopAnnotation(state.get_current_instruction()["address"])
+                        )
+                    else:
```
Comment (norhh) [inline mythril/analysis/modules/dos.py]: shouldn't an addition to jumpdest_count happen above?

File: mythril/analysis/modules/dos.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            if len(annotations) > 0:
+
+                operation = (
+                    "A storage modification" if opcode == "CALL" else "An external call"
+                )
+
+                description_head = (
+                    "Potential Denial-of-Service if block gas limit is reached."
+                )
+                description_tail = operation + " is executed in a loop."
```
Comment (norhh) [inline mythril/analysis/modules/dos.py]: string formatting might be cool here.

File: mythril/analysis/modules/dos.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        else:
+
+            annotations = cast(
+                List[LoopAnnotation], list(state.get_annotations(LoopAnnotation))
+            )
+
+            if len(annotations) > 0:
+
+                operation = (
+                    "A storage modification" if opcode == "CALL" else "An external call"
```
Comment (norhh) [inline mythril/analysis/modules/dos.py]: shouldn't it be reverse?

**Comment by norhh** [comment]
This also detects the SSTORES after the loop right? as the annotations aren't being flushed, one way to deal with it is to check the count of those operation addresses in the loop too.

**Comment by norhh** [comment]
I mean it also shows that in cases like these.
```


contract DosGas {

    address[] creditorAddresses;
    bool win = false;

    function emptyCreditors() public {
        if(creditorAddresses.length>1500) {
            creditorAddresses = new address[](0);
            win = true;
        }
    }

    function addCreditors() public returns (bool) {
        uint x=0;
        for(uint i=0;i<4;i++) {
        }
        creditorAddresses.push(msg.sender);
        return true;
    }

    function iWin() public view returns (bool) {
        return win;
    }

    function numberCreditors() public view returns (uint) {
        return creditorAddresses.length;
    }
}
```

File: mythril/analysis/modules/dos.py (hunk 5 above)
Comment (b-mueller) [reply mythril/analysis/modules/dos.py]: Ammm actually yes :)

**Comment by b-mueller** [comment]
> I mean it also shows that in cases like these.

Should work now I think...

**Pull request merged**